File: code_cpp/self pased DSA/Graph/LEETCODE/5-Shortes-Path-with-Alternating-Colors.cpp
/*
Shortest Path with Alternating Colors
You are given an integer n, the number of nodes in a directed graph where the nodes are labeled from 0 to n - 1. Each edge is red or blue in this graph, and there could be self-edges and parallel edges.

You are given two arrays redEdges and blueEdges where:

redEdges[i] = [ai, bi] indicates that there is a directed red edge from node ai to node bi in the graph, and
blueEdges[j] = [uj, vj] indicates that there is a directed blue edge from node uj to node vj in the graph.
Return an array answer of length n, where each answer[x] is the length of the shortest path from node 0 to node x such that the edge colors alternate along the path, or -1 if such a path does not exist.
*/

class Solution {
public:
    vector<int> shortestAlternatingPaths(int n, vector<vector<int>>& redEdges, vector<vector<int>>& blueEdges) {
        vector<int> dist(n, -1);
        vector<pair<int,int>> adj[n+1];

        for(auto vec: redEdges){
            adj[vec[0]].push_back({vec[1],1}); // source --> [dst, egdeColour], [dst, edgeColor]
        }
        for(auto vec: blueEdges){
            adj[vec[0]].push_back({vec[1],2}); // source --> [dst, egdeColour], [dst, edgeColor]
        }

        queue<vector<int>> q;
        q.push({0, 0, -1}); // <current node, distance from 0 to current node, edgeColor used>

        while(!q.empty()){
            auto currVec = q.front();
            q.pop();

            int currNode = currVec[0], edgeColor=currVec[2];
            if(dist[currNode]==-1){ // node not visited till now
                dist[currNode] = currVec[1];
            } 

            for(auto &childPair: adj[currNode]){
                int childNode = childPair.first;
                int edgeColorOfChild = childPair.second;
                // check if child node is not visited and its color is not same as its parent node edge
                // edgeColor==-1 --> means node 0 have no color
                if(childNode != -1 && (edgeColor==-1 || edgeColor!=edgeColorOfChild)){
                    q.push({childNode, currVec[1]+1, edgeColorOfChild});
                    childPair.first = -1; // mark node is visited
                }
            }
        }
        return dist;
    }
};

File: code_java/self pased DSA/Graph/LEETCODE/5-Shortes-Path-with-Alternating-Colors.java
/*
Problem: Shortest Path with Alternating Colors
Given a directed graph with red and blue edges, find the shortest path from node 0 to every other node such that the edge colors alternate along the path. Return -1 if a node is unreachable.

Approach:
- Build an adjacency list for both red and blue edges, marking edge colors.
- Use BFS starting from node 0, tracking both the current node and the color of the last edge used.
- For each node, only proceed if the next edge color is different from the previous one.
- Mark nodes as visited per color to avoid revisiting with the same color.
- Return the shortest distances for all nodes.
*/

import java.util.*;

public class ShortestPathWithAlternatingColors {
    public int[] shortestAlternatingPaths(int n, int[][] redEdges, int[][] blueEdges) {
        List<int[]>[] adj = new ArrayList[n];
        for (int i = 0; i < n; i++) adj[i] = new ArrayList<>();
        for (int[] edge : redEdges) adj[edge[0]].add(new int[]{edge[1], 1}); // 1 = red
        for (int[] edge : blueEdges) adj[edge[0]].add(new int[]{edge[1], 2}); // 2 = blue
        int[] dist = new int[n];
        Arrays.fill(dist, -1);
        boolean[][] visited = new boolean[n][3]; // visited[node][color]
        Queue<int[]> q = new LinkedList<>();
        q.offer(new int[]{0, 0, -1}); // node, distance, lastColor
        visited[0][0] = true;
        while (!q.isEmpty()) {
            int[] curr = q.poll();
            int node = curr[0], d = curr[1], lastColor = curr[2];
            if (dist[node] == -1) dist[node] = d;
            for (int[] next : adj[node]) {
                int nextNode = next[0], color = next[1];
                if (!visited[nextNode][color] && color != lastColor) {
                    visited[nextNode][color] = true;
                    q.offer(new int[]{nextNode, d + 1, color});
                }
            }
        }
        return dist;
    }
} 